### ISSUE TITLE
TypeError: Converting circular structure to JSON when enabling LangSmith

### ISSUE THREAD

**Issue by sinedied** [langchain-ai/langchainjs#5198]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain.js documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain.js rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

The issue occurs when enabling Langsmith on this sample, in particular the Azure path: https://github.com/Azure-Samples/serverless-chat-langchainjs

The issue seems to come from the `AzureCosmosDBVectorStore`. I'm trying to investigate what could be the cause of the circular structure, do you have any hint on how to debug which object is the root cause?

### Error Message and Stack Trace (if applicable)

```
[api] [2024-04-23T15:11:38.138Z] TypeError: Converting circular structure to JSON
[api] [2024-04-23T15:11:38.138Z]     --> starting at object with constructor 'Socket'
[api] [2024-04-23T15:11:38.138Z]     |     property 'parser' -> object with constructor 'HTTPParser'
[api] [2024-04-23T15:11:38.138Z]     --- property 'socket' closes the circle
[api] [2024-04-23T15:11:38.138Z]     at JSON.stringify (<anonymous>)
[api] [2024-04-23T15:11:38.138Z]     at Client.batchIngestRuns (/Users/sinedied/projects/serverless-chat-langchainjs/node_modules/langsmith/dist/client.cjs:637:51)
[api] [2024-04-23T15:11:38.138Z]     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[api] [2024-04-23T15:11:38.139Z]     at async Client.drainAutoBatchQueue (/Users/sinedied/projects/serverless-chat-langchainjs/node_modules/langsmith/dist/client.cjs:477:17)
```

### Description

Trying to enable Langsmith, but the app crashes when trying to send the report.

### System Info

npm ls langchain @langchain/comm
unity
serverless-chat-langchainjs@1.0.0 /Users/sinedied/projects/serverless-chat-langchainjs
└─┬ api@1.0.0 -> ./packages/api
  ├── @langchain/community@0.0.51 -> ./packages/api
  └─┬ langchain@0.1.35
    └── @langchain/community@0.0.51 deduped -> ./packages/api
    
platform mac, node v20.11.0, npm 10.2.4

**Comment by sinedied** [comment]
Ok, removing entirely `AzureCosmosDBVectorStore` from the chain did not solve the issue. I still get that error, but it cannot be reproduced all the time. Not sure how I should go to debug this or help debugging this?

**Comment by Kniggishood** [comment]
Heyhey, can you provide the code sample for this? What does your wrappedFunc look like?

**Comment by sinedied** [comment]
Yes, I could reproduce the error on this sample/branch: https://github.com/Azure-Samples/serverless-chat-langchainjs/tree/mongodb-vcore

I'm currently off till next week, I'll try to create a simpler repo for reproducing the issue.

**Comment by pnedelko** [comment]
Faced the same issue when trying to implement my custom cache Class.

TLDR: with Langsmith tracing I get error, without tracings works well.

The error:
```
Error in handler LangChainTracer, handleChainEnd: TypeError: Converting circular structure to JSON
     --> starting at object with constructor 'ni'
     |     property 'client' -> object with constructor 'PrismaService'
     --- property '_requestHandler' closes the circle
```

The problem appears when there is a class prop with PrismaClient
```
class PrismaCache extends BaseCache {
   constructor(
     private prisma: PrismaClient,
     private options: { ttl: number },
   ) {
     super();
   }
...
}
```

Temporal workaround: pass an object to the constructor that holds a reference to the PrismaClient in the closure.
```js
{
      get: async (key: string) => {
        const entry = await this.prisma.llmCache.findUnique({
          where: {
            key,
            expiresAt: {
              gt: new Date(),
            },
          },
        });

        if (!entry) {
          return null;
        }

        return entry.value;
      },
      set: async (key: string, value: any, ttl: number) => {
        await this.prisma.llmCache.upsert({
          where: {
            key,
          },
          create: {
            key,
            value: value as any,
            expiresAt: new Date(Date.now() + ttl),
          },
          update: {
            value: value as any,
            expiresAt: new Date(Date.now() + ttl),
          },
        });
      },
    }
```

**Comment by airhorns** [comment]
I'm experiencing this issue as well -- it seems like Langsmith tries to blindly serialize the arguments and return values of traced functions, and if those values happen to be rich objects that reference one another, they can't be safely serialized like that. Could we get an API added to allow a transform of the values before being serialized? Some of the objects that my functions use are also quite big if serialized -- maybe it's the wrong default to serialize everything blindly?

**Comment by airhorns** [comment]
Another option would be to use one of the circular-safe serialization libraries

**Comment by mbennett-talentnet** [comment]
Any update on this? I have enabled MongoDBSaver for checkpointing in langgraph and now langsmith is broken with the following error:

```
TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'MongoClient'
    |     property 's' -> object with constructor 'Object'
    |     property 'sessionPool' -> object with constructor 'ServerSessionPool'
    --- property 'client' closes the circle
    at JSON.stringify (<anonymous>)
    at Client.batchIngestRuns (file:///var/www/app/node_modules/langsmith/dist/client.js:612:51)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Client.drainAutoBatchQueue (file:///var/www/app/node_modules/langsmith/dist/client.js:451:17)
```

**Comment by jacoblee93** [comment]
Thanks for reporting - will identify and patch as soon as possible!

**Comment by jacoblee93** [comment]
PR opened here: https://github.com/langchain-ai/langsmith-sdk/pull/985

**Comment by jacoblee93** [comment]
This should be fixed although in the latest LangSmith version I'd still recommend avoiding circular references for performance issues.
